### PR TITLE
Removes redundant copy in ExternalSource operator

### PR DESCRIPTION
**Pull request by JanuszL**
- removes the redundant copy from CPU and GPU backend - now ExternalSource will just use data from staging buffer directly inside RunImpl
- adds NVTX range to SetDataSource so now it is possible to measure user thread time of feeding ExternalSource operator
- adds a more verbose message when operator and data backend doesn't match when zero_copy is used
- fixes lack of waiting for the inputs when zero-copy is set
- adds Swap method to TensorVector and TensorList
- ExternalSource doesn't always sync anymore when the pinned CPU buffer is provided for the GPU operator

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It fixes a lack of waiting for the inputs when zero-copy is set
- It removes the redundant copy from CPU and GPU backend - now ExternalSource will just use data from staging buffer directly inside RunImpl

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     removes the redundant copy from CPU backend - now ExternalSource will just use data from staging buffer directly inside RunImpl
     adds NVTX range to SetDataSource so now it is possible to measure user thread time of feeding ExternalSource operator
     adds a more verbose message when operator and data backend doesn't match when zero_copy is used
     fixes lack of waiting for the inputs when zero-copy is set
     ExternalSource doesn't always sync anymore when the pinned CPU buffer is provided for the GPU operator
 - Affected modules and functionalities:
     ExternalSource
     TensorList
     TensorVector
     Buffer
 - Key points relevant for the review:
     NA
 - Validation and testing:
     CI
 - Documentation (including examples):
     NA

**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/pipeline/data/tensor_vector.h [hunk 1]
```diff
@@ -65,6 +65,21 @@ class TensorVector {
...
 
+  DLL_PUBLIC TensorVector<Backend>(TensorVector<Backend> &&other) noexcept {
+    state_ = other.state_;
+    pinned_ = other.pinned_;
+    tl_ = std::move(other.tl_);
+    type_ = other.type_;
+    views_count_ = other.views_count_.load();
+    tensors_ = std::move(other.tensors_);
+    for (auto &t : tensors_) {
+      if (auto *del = std::get_deleter<ViewRefDeleter>(t->data_)) del->ref = &views_count_;
```
Comment (mzient) [inline dali/pipeline/data/tensor_vector.h:76]: I'm not sure if it's possible for `tensors_` to contain nullptrs, but if it is, then
```suggestion
      if (t)
        if (auto *del = std::get_deleter<ViewRefDeleter>(t->data_))
          del->ref = &views_count_;
```

File: dali/pipeline/data/tensor_vector.h [hunk 2]
```diff
@@ -375,6 +390,24 @@ class TensorVector {
...
+  inline TensorVector<Backend>& operator=(TensorVector<Backend> &&other) noexcept {
+    if (&other != this) {
+      state_ = other.state_;
+      pinned_ = other.pinned_;
+      tl_ = std::move(other.tl_);
+      type_ = other.type_;
+      views_count_ = other.views_count_.load();
+      tensors_ = std::move(other.tensors_);
+      for (auto &t : tensors_) {
+        if (auto *del = std::get_deleter<ViewRefDeleter>(t->data_)) del->ref = &views_count_;
```
Comment (mzient) [inline dali/pipeline/data/tensor_vector.h:402]: likewise

Comment (JanuszL) [reply dali/pipeline/data/tensor_vector.h]: Done

File: dali/pipeline/data/tensor_vector.h (hunk 1 above)
Comment (JanuszL) [reply dali/pipeline/data/tensor_vector.h]: Done

**Comment by JanuszL** [comment]
!build

**Pull request merged**